src/main/java/io/javarig/TypeEnum.java

```
@@ -99,17 +91,12 @@ private static TypeEnum getObjectIfNotEnum(Type type) {
         return OBJECT;
     }
 
-    private void setRandomInstanceGenerator(RandomInstanceGenerator randomInstanceGenerator) {
-        this.generator.setRandomInstanceGenerator(randomInstanceGenerator);
-    }
-
-    public TypeGenerator generator() {
-        return this.generator;
-    }
-
-    public void setType(Type type) {
-        if (this.generator instanceof TypeBasedGenerator typeBasedGenerator) {
-            typeBasedGenerator.setType(type);
+    public TypeGenerator createGeneratorInstance(Type type, RandomInstanceGenerator randomInstanceGenerator) {
+        try {
+            return (AbstractTypeGenerator) this.generatorClass.getConstructor(Type.class, RandomInstanceGenerator.class)
+                    .newInstance(type, randomInstanceGenerator);
+        } catch (ReflectiveOperationException e) {
```

i think its better to throw a custom exception instead!